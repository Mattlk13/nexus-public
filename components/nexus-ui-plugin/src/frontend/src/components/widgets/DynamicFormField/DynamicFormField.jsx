/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import React from 'react';
import {mapObjIndexed, values} from 'ramda';
import Textfield from "../Textfield/Textfield";
import MultiSelect from '../MultiSelect/MultiSelect';
import Select from '../Select/Select';
import Utils from '../../../interface/Utils';

export default function DynamicFormField({current, dynamicProps, id, initialValue, onChange}) {
  if (dynamicProps.type === 'string') {
    const fieldProps = Utils.fieldProps(id, current, initialValue || '');
    const className = dynamicProps.attributes.long ? 'nx-text-input--long' : '';

    return <Textfield {...fieldProps}
                      className={className}
                      disabled={dynamicProps.disabled}
                      readOnly={dynamicProps.readOnly}
                      onChange={(event) => onChange(fieldProps.name, event.currentTarget.value)}
    />
  }
  else if (dynamicProps.type === 'itemselect') {
    const fieldProps = Utils.fieldProps(id, current, initialValue || [], (value) => value)
    return <MultiSelect
        {...fieldProps}
        filterValue={dynamicProps.filterValue}
        toLabel={dynamicProps.attributes.toTitle}
        fromLabel={dynamicProps.attributes.fromTitle}
        fromOptions={dynamicProps.attributes.options}
        onChange={(value) => onChange(fieldProps.name, value)}
    />;
  }
  else if (dynamicProps.type === 'combobox') {
    const fieldProps = Utils.fieldProps(id, current, initialValue || '');
    return <Select {...fieldProps} onChange={(event) => onChange(fieldProps.name, event.currentTarget.value)}>
      <option/>
      {values(mapObjIndexed((v, k) => <option key={k} value={k}>{v}</option>, dynamicProps.attributes.options))}
    </Select>;
  }
  else {
    console.warn(`form field type=${dynamicProps.type} is unknown`);
    return <div/>;
  }
}
