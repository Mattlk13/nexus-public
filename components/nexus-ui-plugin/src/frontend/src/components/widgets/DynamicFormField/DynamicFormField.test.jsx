/*
 * Sonatype Nexus (TM) Open Source Version
 * Copyright (c) 2008-present Sonatype, Inc.
 * All rights reserved. Includes the third-party code listed at http://links.sonatype.com/products/nexus/oss/attributions.
 *
 * This program and the accompanying materials are made available under the terms of the Eclipse Public License Version 1.0,
 * which accompanies this distribution and is available at http://www.eclipse.org/legal/epl-v10.html.
 *
 * Sonatype Nexus (TM) Professional Version is available from Sonatype, Inc. "Sonatype" and "Sonatype Nexus" are trademarks
 * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
 * Eclipse Foundation. All other trademarks are the property of their respective owners.
 */
import React from 'react';
import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import DynamicFormField from "./DynamicFormField";

describe('DynamicFormField', () => {
  describe('renders the string type', () => {
    const stringFieldProps = {
      id: 'test',
      current: {
        context: {
          data: {
            test: ''
          },
          pristineData: {
            test: ''
          }
        }
      }
    };

    it('renders', () => {
      const {container} = render(<DynamicFormField {...stringFieldProps} dynamicProps={{
        type: 'string',
        attributes: {}
      }}/>);

      expect(container).toMatchSnapshot();
    });

    it('renders a long field', () => {
      const {container} = render(<DynamicFormField {...stringFieldProps} dynamicProps={{
        type: 'string',
        attributes: {long: true}
      }}/>);

      expect(container).toMatchSnapshot();
    });

    it('renders a disabled field', () => {
      const {getByRole} = render(<DynamicFormField {...stringFieldProps} dynamicProps={{
        type: 'string',
        attributes: {},
        disabled: true
      }}/>);

      expect(getByRole('textbox')).toBeDisabled();
    });

    it('renders a readonly field', () => {
      const {getByRole} = render(<DynamicFormField {...stringFieldProps} dynamicProps={{
        type: 'string',
        attributes: {},
        readOnly: true
      }}/>);

      expect(getByRole('textbox')).toHaveAttribute('readonly');
    });

    it('renders a MultiSelect field', () => {
      const {container} = render(<DynamicFormField {...makeContext('test', ['a'])} dynamicProps={{
        type: 'itemselect',
        attributes: {
          options: ['a', 'b']
        }
      }}/>);

      expect(container).toMatchSnapshot();
    });

    it('renders a Select field', () => {
      const {container} = render(<DynamicFormField {...makeContext('test', 'a')} dynamicProps={{
        type: 'combobox',
        attributes: {
          options: {'a': 'a', 'b':'b'}
        }
      }}/>);

      expect(container).toMatchSnapshot();
    });

    function makeContext(id, value) {
      return {
        id,
        current: {
          context: {
            data: {
              [id]: value
            },
            pristineData: {
              [id]: value
            }
          }
        }
      }
    }
  });
});
