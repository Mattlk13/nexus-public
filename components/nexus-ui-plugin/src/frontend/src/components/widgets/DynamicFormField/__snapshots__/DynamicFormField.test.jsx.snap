// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicFormField renders the string type renders 1`] = `
<div>
  <div
    class="nx-text-input pristine"
  >
    <div
      class="nx-text-input__box"
    >
      <input
        aria-errormessage="nx-text-input-invalid-message-00000000"
        aria-invalid="false"
        class="nx-text-input__input"
        id="test"
        name="test"
        type="text"
        value=""
      />
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check fa-w-16 nx-icon nx-icon--valid nx-icon"
        data-icon="check"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-exclamation-circle fa-w-16 nx-icon nx-icon--invalid nx-icon"
        data-icon="exclamation-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="nx-text-input__invalid-message"
      id="nx-text-input-invalid-message-00000000"
      role="alert"
    />
  </div>
</div>
`;

exports[`DynamicFormField renders the string type renders a MultiSelect field 1`] = `
<div>
  <div
    class="nxrm-multiselect"
  >
    <div
      class="nxrm-multiselect__from"
    >
      <label
        class="nx-sub-label"
        for="nxrm-multiselect-from-44444444"
      >
        Available
      </label>
      <div
        class="nx-text-input nx-filter-input"
      >
        <div
          class="nx-text-input__box"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-filter fa-w-16 nx-icon--filter-icons nx-icon"
            data-icon="filter"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
              fill="currentColor"
            />
          </svg>
          <input
            aria-errormessage="nx-text-input-invalid-message-66666666"
            class="nx-text-input__input"
            type="text"
            value=""
          />
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-check fa-w-16 nx-icon nx-icon--valid nx-icon"
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-exclamation-circle fa-w-16 nx-icon nx-icon--invalid nx-icon"
            data-icon="exclamation-circle"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
              fill="currentColor"
            />
          </svg>
        </div>
        <div
          class="nx-text-input__invalid-message"
          id="nx-text-input-invalid-message-66666666"
          role="alert"
        />
      </div>
      <div
        class="nxrm-select"
      >
        <select
          class="nx-form-select"
          id="nxrm-multiselect-from-44444444"
          multiple=""
        >
          <option
            value="b"
          >
            b
          </option>
        </select>
      </div>
    </div>
    <div
      class="nxrm-multiselect__controls"
    >
      <button
        aria-label="Move Selection Right"
        class="nx-btn nx-btn--icon-only"
        title="Move Selection Right"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-right fa-w-6 nx-icon"
          data-icon="caret-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 192 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Move Selection Left"
        class="nx-btn nx-btn--icon-only"
        title="Move Selection Left"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-left fa-w-6 nx-icon"
          data-icon="caret-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 192 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M192 127.338v257.324c0 17.818-21.543 26.741-34.142 14.142L29.196 270.142c-7.81-7.81-7.81-20.474 0-28.284l128.662-128.662c12.599-12.6 34.142-3.676 34.142 14.142z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
    <div
      class="nxrm-multiselect__to"
    >
      <label
        class="nx-sub-label"
        for="nxrm-multiselect-to-55555555"
      >
        Selected
      </label>
      <div
        class="nxrm-select"
      >
        <select
          class="nx-form-select"
          id="nxrm-multiselect-to-55555555"
          multiple=""
        >
          <option
            value="a"
          >
            a
          </option>
        </select>
      </div>
    </div>
    <div
      class="nxrm-multiselect__controls"
    >
      <button
        aria-label="Move Selection Up"
        class="nx-btn nx-btn--icon-only"
        title="Move Selection Up"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-up fa-w-10 nx-icon"
          data-icon="caret-up"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M288.662 352H31.338c-17.818 0-26.741-21.543-14.142-34.142l128.662-128.662c7.81-7.81 20.474-7.81 28.284 0l128.662 128.662c12.6 12.599 3.676 34.142-14.142 34.142z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label="Move Selection Down"
        class="nx-btn nx-btn--icon-only"
        title="Move Selection Down"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-caret-down fa-w-10 nx-icon"
          data-icon="caret-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`DynamicFormField renders the string type renders a Select field 1`] = `
<div>
  <div
    class="nxrm-select"
  >
    <select
      class="nx-form-select"
      id="test"
      name="test"
    >
      <option />
      <option
        value="a"
      >
        a
      </option>
      <option
        value="b"
      >
        b
      </option>
    </select>
  </div>
</div>
`;

exports[`DynamicFormField renders the string type renders a long field 1`] = `
<div>
  <div
    class="nx-text-input nx-text-input--long pristine"
  >
    <div
      class="nx-text-input__box"
    >
      <input
        aria-errormessage="nx-text-input-invalid-message-11111111"
        aria-invalid="false"
        class="nx-text-input__input"
        id="test"
        name="test"
        type="text"
        value=""
      />
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-check fa-w-16 nx-icon nx-icon--valid nx-icon"
        data-icon="check"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-exclamation-circle fa-w-16 nx-icon nx-icon--invalid nx-icon"
        data-icon="exclamation-circle"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zm-248 50c-25.405 0-46 20.595-46 46s20.595 46 46 46 46-20.595 46-46-20.595-46-46-46zm-43.673-165.346l7.418 136c.347 6.364 5.609 11.346 11.982 11.346h48.546c6.373 0 11.635-4.982 11.982-11.346l7.418-136c.375-6.874-5.098-12.654-11.982-12.654h-63.383c-6.884 0-12.356 5.78-11.981 12.654z"
          fill="currentColor"
        />
      </svg>
    </div>
    <div
      class="nx-text-input__invalid-message"
      id="nx-text-input-invalid-message-11111111"
      role="alert"
    />
  </div>
</div>
`;
